matrix-sdk/src/main/java/org/matrix/androidsdk/login/AutoDiscovery.kt

```diff
@@ -0,0 +1,246 @@
+import java.io.IOException
+import java.net.URL
+
+
+class AutoDiscovery(var okHttpClient: OkHttpClient = OkHttpClient()) {
+
+    companion object {
+        val instance = AutoDiscovery()
+        private val LOG_TAG = AutoDiscovery::class.java.name
+    }
+
+    data class DiscoveredClientConfig(val action: Action, val homeServerUrl: String? = null, val idendityServerUrl: String? = null)
+
+    enum class Action {
+        /*
+         Retrieve the specific piece of information from the user in a way which fits within the existing client user experience,
+         if the client is inclined to do so. Failure can take place instead if no good user experience for this is possible at this point.
+         */
+        PROMPT,
+        /*
+         Stop the current auto-discovery mechanism. If no more auto-discovery mechanisms are available,
+         then the client may use other methods of determining the required parameters, such as prompting the user, or using default values.
+         */
+        IGNORE,
+        /*
+            Inform the user that auto-discovery failed due to invalid/empty data and PROMPT for the parameter.
+         */
+        FAIL_PROMPT,
+        /*
+         Inform the user that auto-discovery did not return any usable URLs. Do not continue further with the current login process.
+         At this point, valid data was obtained, but no homeserver is available to serve the client.
+         No further guess should be attempted and the user should make a conscientious decision what to do next.
+         */
+        FAIL_ERROR,
+
+    }
+
+    private val mUiHandler = Handler(Looper.getMainLooper());
+
+    fun findClientConfig(domain: String, callback: ApiCallback<DiscoveredClientConfig>) {
+
+        val request = Request.Builder()
+                .url("https://$domain/.well-known/matrix/client")
+                .build()
+
+        /**
+         * Fetches a JSON object from a given URL, as expected by all .well-known
+         * related lookups. If the server gives a 404 then the `action` will be
+         * IGNORE. If the server returns something that isn't JSON, the `action`
+         * will be FAIL_PROMPT. For any other failure the `action` will be FAIL_PROMPT.
+         *
+         * The returned object will be a result of the call in object form with
+         * the following properties:
+         *   raw: The JSON object returned by the server.
+         *   action: One of SUCCESS, IGNORE, or FAIL_PROMPT.
+         *   reason: Relatively human readable description of what went wrong.
+         *   error: The actual Error, if one exists.
+         *   */
+        okHttpClient.newCall(request).enqueue(object : Callback {
+
+            override fun onResponse(call: Call, response: Response) {
+                val code = response.code()
+                //a. If the returned status code is 404, then IGNORE.
+                if (code == 404) {
+                    //IGNORE
+                    mUiHandler.post {
```

Comment:
no need to switch to UI thread here (and in other cases also)

Reply:
Was not the case with okhttp, but ok with retrofit